File: crates/graphql/src/model/anki.rs
#[derive(async_graphql::SimpleObject)]
pub struct AnkiConnection {
    pub edges: Vec<AnkiEdge>,
    pub page_info: crate::model::PageInfo,
}

#[derive(async_graphql::SimpleObject)]
pub struct AnkiEdge {
    pub node: Anki,
    pub cursor: String,
}

pub struct Anki {
    pub page_id: String,
    pub title: Option<String>,
    pub description: Option<String>,
    pub ease_factor: f64,
    pub repetition_count: u32,
    pub next_review_at: String,
    pub created_at: String,
    pub updated_at: String,
    pub tags: Vec<AnkiTag>,
    pub url: String,
}

#[derive(async_graphql::SimpleObject)]
pub struct AnkiTag {
    pub id: String,
    pub name: String,
    pub color: String,
}

#[derive(async_graphql::SimpleObject)]
pub struct AnkiBlock {
    pub front: serde_json::Value,
    pub back: serde_json::Value,
    pub explanation: serde_json::Value,
}

#[async_graphql::Object]
impl Anki {
    async fn page_id(&self) -> &str {
        &self.page_id
    }

    async fn title(&self) -> Option<&str> {
        self.title.as_deref()
    }

    async fn description(&self) -> Option<&str> {
        self.description.as_deref()
    }

    async fn ease_factor(&self) -> f64 {
        self.ease_factor
    }

    async fn repetition_count(&self) -> u32 {
        self.repetition_count
    }

    async fn next_review_at(&self) -> &str {
        &self.next_review_at
    }

    async fn created_at(&self) -> &str {
        &self.created_at
    }

    async fn updated_at(&self) -> &str {
        &self.updated_at
    }

    async fn tags(&self) -> &[AnkiTag] {
        &self.tags
    }

    async fn url(&self) -> &str {
        &self.url
    }

    pub async fn block_list(&self) -> Result<AnkiBlock, async_graphql::Error> {
        let secret = std::env::var("NOTION_API_KEY")
            .map_err(|_| async_graphql::Error::from("NOTION_API_KEY not found"))?;

        let mut client = elmethis_notion::client::Client::new(secret);

        let blocks = client
            .convert_block(&self.page_id)
            .await
            .map_err(|e| async_graphql::Error::from(format!("Failed to get block: {}", e)))?;

        let mut front: Vec<elmethis_notion::block::Block> = Vec::new();
        let mut back: Vec<elmethis_notion::block::Block> = Vec::new();
        let mut explanation: Vec<elmethis_notion::block::Block> = Vec::new();

        enum Marker {
            Front,
            Back,
            Explanation,
        }

        let mut marker = Marker::Front;

        for block in blocks {
            if let elmethis_notion::block::Block::ElmHeading1(
                elmethis_notion::block::ElmHeading1 { props },
            ) = &block
            {
                if props.text == "front" {
                    marker = Marker::Front;
                    continue;
                } else if props.text == "back" {
                    marker = Marker::Back;
                    continue;
                } else if props.text == "explanation" {
                    marker = Marker::Explanation;
                    continue;
                }
            }

            match marker {
                Marker::Front => front.push(block),
                Marker::Back => back.push(block),
                Marker::Explanation => explanation.push(block),
            }
        }

        Ok(AnkiBlock {
            front: serde_json::to_value(front)?,
            back: serde_json::to_value(back)?,
            explanation: serde_json::to_value(explanation)?,
        })
    }
}

File: packages/http-api/server/features/anki.ts
import type {
  BlockObjectRequest,
  PageObjectResponse,
  UpdatePageParameters,
} from "@notionhq/client";
import { Surface } from "n2a2ui";
import type { z } from "zod";
import type { schemas } from "../contracts.ts";
import type { NotionRepository } from "../lib/notion.ts";
import {
  property,
  required,
  richText,
  tagColors,
  text,
  unsigned,
} from "../lib/properties.ts";

export function ankiEntity(page: PageObjectResponse) {
  return {
    page_id: page.id,
    title: richText(property(page, "title", "title").title).trim() || null,
    description:
      richText(property(page, "description", "rich_text").rich_text).trim() ||
      null,
    ease_factor: required(
      property(page, "easeFactor", "number").number,
      "easeFactor",
    ),
    repetition_count: unsigned(
      required(
        property(page, "repetitionCount", "number").number,
        "repetitionCount",
      ),
    ),
    next_review_at: required(
      property(page, "nextReviewAt", "date").date,
      "nextReviewAt",
    ).start,
    created_at: page.created_time,
    updated_at: page.last_edited_time,
    tags: property(page, "tags", "multi_select").multi_select.map((tag) => ({
      ...tag,
      color: tagColors[tag.color],
    })),
    url: page.url,
    is_review_required: property(page, "isReviewRequired", "checkbox").checkbox,
  };
}

export function splitSections(surface: Surface) {
  const sections: Record<"front" | "back" | "explanation", string[]> = {
    front: [],
    back: [],
    explanation: [],
  };
  let section: keyof typeof sections = "front";
  const root = surface.components.get(surface.root);
  const children =
    root?.component === "Column" && Array.isArray(root.children)
      ? root.children
      : [];
  for (const id of children) {
    const component = surface.components.get(id);
    if (
      component?.component === "Heading" &&
      component.level === 1 &&
      Array.isArray(component.children)
    ) {
      const marker = component.children
        .map((childId) => {
          const child = surface.components.get(childId);
          return child?.component === "RichText" &&
            typeof child.text === "string"
            ? child.text
            : "";
        })
        .join("")
        .trim()
        .toLowerCase();
      if (marker === "front" || marker === "back" || marker === "explanation") {
        section = marker;
        continue;
      }
    }
    sections[section].push(id);
  }
  const build = (ids: string[]) => {
    const result = new Surface("root", surface.components);
    result.components.delete(surface.root);
    result.insert({ id: "root", component: "Column", children: ids });
    return result.toJSON();
  };
  return {
    front: build(sections.front),
    back: build(sections.back),
    explanation: build(sections.explanation),
  };
}

export class AnkiService {
  constructor(readonly repository: NotionRepository) {}
  async list(pageSize: number, cursor?: string) {
    const { pages } = await this.repository.query("anki", {
      page_size: pageSize,
      start_cursor: cursor,
      sorts: [{ property: "nextReviewAt", direction: "ascending" }],
    });
    return pages.map(ankiEntity);
  }
  async get(id: string) {
    return ankiEntity(await this.repository.getPage(id));
  }
  async blocks(id: string) {
    return splitSections(await this.repository.converter.convertBlock(id));
  }
  async create(input: z.infer<typeof schemas.CreateAnkiRequest>) {
    const children: BlockObjectRequest[] = [
      "front",
      "back",
      "explanation",
    ].flatMap((section) => [
      {
        object: "block" as const,
        type: "heading_1" as const,
        heading_1: {
          rich_text: [
            { ...text(section)[0], annotations: { color: "brown" as const } },
          ],
        },
      },
      {
        object: "block" as const,
        type: "paragraph" as const,
        paragraph: { rich_text: [] },
      },
    ]);
    return ankiEntity(
      await this.repository.createPage("anki", {
        properties: {
          title: { title: text(input.title ?? "No Title") },
          easeFactor: { number: 2.5 },
          repetitionCount: { number: 0 },
          nextReviewAt: { date: { start: new Date().toISOString() } },
        },
        children,
      }),
    );
  }
  async update(id: string, input: z.infer<typeof schemas.UpdateAnkiRequest>) {
    const properties: NonNullable<UpdatePageParameters["properties"]> = {};
    if (input.ease_factor != null)
      properties.easeFactor = { number: input.ease_factor };
    if (input.repetition_count != null)
      properties.repetitionCount = { number: input.repetition_count };
    if (input.next_review_at != null)
      properties.nextReviewAt = { date: { start: input.next_review_at } };
    if (input.is_review_required != null)
      properties.isReviewRequired = { checkbox: input.is_review_required };
    return ankiEntity(
      await this.repository.updatePage(id, {
        properties,
        in_trash: input.in_trash ?? undefined,
      }),
    );
  }
}
